openml/extensions/sklearn/extension.py

```diff
@@ -0,0 +1,134 @@
+from typing import Any, Dict, List, Optional, Tuple, TYPE_CHECKING
+
+import sklearn.base
+
+from openml.extensions import Extension, register_extension
+from openml.extensions.sklearn.flow_functions import (
```

Comment:
why put these functions in different packages? why not implement them in this file?

Reply:
To keep the file rather small. Also, they're actually taking care of different things, namely serializing/deserializing flows and running flows.

Reply:
Do you think it might make sense to use double inheritance instead? E.g. basically make all functions in `run_functions.py` part of a `SklearnRunExtension`-class and all functions in `flow_functions` part of a `SklearnFlowExtension`-class and then simply use inheritance instead (both of these would inherit from `Extension`). Functions which are implemented in neither can be implemented here (currently only `can_handle_model`). So the rewritten extension would look like:

```
class SklearnExtension(SklearnRunExtension, SklearnFlowExtension):

    @staticmethod
    def can_handle_model(model: Any) -> bool:
        """Check whether a model is an instance of ``sklearn.base.BaseEstimator``.

        Parameters
        ----------
        model : Any

        Returns
        -------
        bool
        """
        return isinstance(model, sklearn.base.BaseEstimator)
```
?

I don't have much experience with multiple inheritance but I think it would work (I tried a toy example and that worked). It would help by:
 - Removing a lot of double documentation, since now you document the function twice, e.g. [here](https://github.com/openml/openml-python/blob/extensions/openml/extensions/sklearn/extension.py#L228) and [here](https://github.com/openml/openml-python/blob/extensions/openml/extensions/sklearn/flow_functions.py#L262). This can lead to inconsistencies (indeed, even the linked example has inconsistencies, though only a minor ones regarding spacing.
 - making it much clearer which functions are actually implemented in this file.
 - maintenance when e.g. function signatures change. Currently we should change the `Extension` interface, the method implementation in e.g. `flow_functions.py` and the redirect in `extension.py`. Using multiple inheritance saves us one step (since the method is not redeclared in `extension.py`)

Reply:
To add to this. 

I see how it might be sub-optimal to handle this in the current PR, as this increases the size of the diff significantly, but this is something that we should have resolved before making a new release.

Reply:
I can see your points, especially regarding double documentation, and clarity about where the functions are.

I'll move all code into the class. This shouldn't make this PR bigger as it does not make a difference to the total PR size where the code is copied to.

Reply:
I don't currently foresee the problems that Jan mentioned could be an issue, but that is probably  due a lack of working with that particular code. Either way I trust the judgement, and the adjustments Jan proposed seem fine with me otherwise too.

Reply:
The idea was to still have the distinction between run/flow operations through two classes (and for me, preferably in two files). Now everything is back in one monolithic file/class.
I think Jan's version of the proposal was that there are two classes:
`class SklearnFlowExtension(Extension):` and `class SklearnExtension(SklearnFlowExtension)`. Here, the former implements the functions previously in `flow_functions` and the latter implements the functions previously in `run_functions` and `sklearn\extension.py`. Having the structure clear both in classes and files makes it easier to navigate the code base, has less chance for merge conflicts, and more explicitly defines the assumed dependencies.

Is there are reason you went with putting it all in one class/file, or is it a misunderstanding (possibly on my part)?

Reply:
I think the original proposal by @janvanrijn was to implement all functions related to the sklearn extension in one file. Your proposal with double inheritance would solve the issue with duplicated docstrings, but in my opinion would not help to make it clearer where the functionality is implemented compared to having two files with functions. Since having everything in a single file appears to be conceptually simpler and also easier maintain for the near future I went with this solution.

Reply:
Hmm. I don't really agree but at least this still solves the double documentation issues. If you prefer this and @janvanrijn has no objections then no further changes are needed in my opinion.